File: inappmessaging/app/src/main/java/com/google/firebase/fiamquickstart/java/MainActivity.java
package com.google.firebase.fiamquickstart.java;

import android.os.Bundle;
import android.util.Log;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;

import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.material.snackbar.Snackbar;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.google.firebase.fiamquickstart.R;
import com.google.firebase.fiamquickstart.databinding.ActivityMainBinding;
import com.google.firebase.inappmessaging.FirebaseInAppMessaging;
import com.google.firebase.installations.FirebaseInstallations;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "FIAM-Quickstart";

    private FirebaseAnalytics mFirebaseAnalytics;
    private FirebaseInAppMessaging mInAppMessaging;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        final ActivityMainBinding binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        mFirebaseAnalytics = FirebaseAnalytics.getInstance(this);
        mInAppMessaging = FirebaseInAppMessaging.getInstance();

        mInAppMessaging.setAutomaticDataCollectionEnabled(true);
        mInAppMessaging.setMessagesSuppressed(false);

        binding.eventTriggerButton.setOnClickListener(
                new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        mFirebaseAnalytics.logEvent("engagement_party", new Bundle());
                        Snackbar.make(view, "'engagement_party' event triggered!", Snackbar.LENGTH_LONG)
                                .setAction("Action", null)
                                .show();
                    }
                });

        // Get and display/log the Instance ID
        FirebaseInstallations.getInstance().getId()
                .addOnSuccessListener(new OnSuccessListener<String>() {
                    @Override
                    public void onSuccess(String id) {
                        binding.installationIdText.setText(getString(R.string.installation_id_fmt, id));
                        Log.d(TAG, "Installation id: " + id);
                    }
                });
    }
}

File: messaging/app/src/main/java/com/google/firebase/quickstart/fcm/kotlin/MainActivity.kt
package com.google.firebase.quickstart.fcm.kotlin

import android.os.Build
import android.os.Bundle
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.quickstart.fcm.R
import com.google.firebase.quickstart.fcm.databinding.ActivityMainBinding
import com.google.firebase.quickstart.fcm.kotlin.data.SubscriptionState
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {
    private val viewModel: FirebaseMessagingViewModel by viewModels { FirebaseMessagingViewModel.Factory }

    private val requestPermissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestPermission()
    ) { isGranted: Boolean ->
        if (isGranted) {
            Toast.makeText(this, "Notifications permission granted", Toast.LENGTH_SHORT)
                .show()
        } else {
            Toast.makeText(this, "FCM can't post notifications without POST_NOTIFICATIONS permission",
                Toast.LENGTH_LONG).show()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            // Create channel to show notifications.
            viewModel.setNotificationChannel(
                this,
                getString(R.string.default_notification_channel_id),
                getString(R.string.default_notification_channel_name)
            )
        }

        // If a notification message is tapped, any data accompanying the notification
        // message is available in the intent extras. In this sample the launcher
        // intent is fired when the notification is tapped, so any accompanying data would
        // be handled here. If you want a different intent fired, set the click_action
        // field of the notification message to the desired intent. The launcher intent
        // is used when no click_action is specified.
        //
        // Handle possible data accompanying notification message.
        // [START handle_data_extras]
        viewModel.logNotificationData(intent)
        // [END handle_data_extras]

        binding.subscribeButton.setOnClickListener {
            // [START subscribe_topics]
            viewModel.subscribeToTopic("weather")
            // [END subscribe_topics]
        }

        binding.logTokenButton.setOnClickListener {
            // [START log_reg_token]
            viewModel.getToken()
            // [END log_reg_token]
        }

        lifecycleScope.launch {
            viewModel.token.collect { token ->
                if(token.isNotEmpty()){
                    val msg = getString(R.string.msg_token_fmt, token)
                    Snackbar.make(findViewById(android.R.id.content),
                        msg, Snackbar.LENGTH_LONG).show()
                }
            }
        }

        lifecycleScope.launch {
            viewModel.subscriptionState.collect { state ->
                when (state) {
                    SubscriptionState.Success -> {
                        Snackbar.make(findViewById(android.R.id.content),
                            getString(R.string.msg_subscribed), Snackbar.LENGTH_LONG).show()
                    }
                    SubscriptionState.Failed -> {
                        Snackbar.make(findViewById(android.R.id.content),
                            getString(R.string.msg_subscribe_failed), Snackbar.LENGTH_LONG).show()
                    }
                    SubscriptionState.Loading -> { }
                }
            }
        }
        Toast.makeText(this, getString(R.string.msg_setup_readme_instructions), Toast.LENGTH_SHORT).show()

        lifecycleScope.launch {
            viewModel.askNotificationPermission(applicationContext, requestPermissionLauncher)
        }
    }

    companion object {
        private const val TAG = "MainActivity"
    }
}
